### ISSUE TITLE
Can't build

### ISSUE THREAD

**Issue by SolitudeSF** [ondrejbudai/hidviz#25]
Build fails at `make` stage with
```
libhidx/libhidx_server/CMakeFiles/hidx_server.dir/build.make:61: *** target pattern contains no '%'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:248: libhidx/libhidx_server/CMakeFiles/hidx_server.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

**Comment by ondrejbudai** [comment]
I need more info - what is your OS?

**Comment by SolitudeSF** [comment]
Void Linux

**Comment by ondrejbudai** [comment]
Unfortunately, I am not able to reproduce, I was able to build it on Void. Have you used the following steps?

```
git clone --recursive https://github.com/ondrejbudai/hidviz.git
mkdir hidviz/build
cd hidviz/build
cmake ..
make
```

**Comment by SolitudeSF** [comment]
yes, exactly that. maybe thats a problem in my setup.

**Comment by SolitudeSF** [comment]
strange, it fails to build with same `make` error even with xbps-src

**Comment by SolitudeSF** [comment]
2 other people on void tried building it and got exact same error. are you sure your void install is up-to-date? this might be a void's issue, but still

**Comment by ondrejbudai** [comment]
I tried it only inside void container which worked weirdly but was able to compile hidviz. So, let's try two things:

1) I will try to test it in VM. Hopefully, I will be able to find an image ready to run. If you can point me at some, I will be glad. :)
2) Can you post me the full verbose output from cmake and make?

**Comment by SolitudeSF** [comment]
here are the outputs from cmake and make with `debug` flags.
http://ix.io/23BJ
http://ix.io/23BK

void-live-* images should work.
https://alpha.de.repo.voidlinux.org/live/20191109/

**Comment by Karrq** [comment]
Coming from Arch's AUR

<details>
<summary> make output </summary>

```
==> Starting prepare()...
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is GNU 11.1.0
-- The CXX compiler identification is GNU 11.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.33.0") 
Cloning into 'asio'...
Note: switching to '8d4c8c3ce43c866f609d2eda9a43fe5b334620be'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

-- Found Protobuf: /usr/lib/libprotobuf.so;-lpthread (found version "3.17.3") 
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Found libusb-1.0:
--  - Includes: /usr/include/libusb-1.0
--  - Libraries: /usr/lib/libusb-1.0.so
-- Configuring done
CMake Warning (dev) in hidviz/CMakeLists.txt:
  Policy CMP0100 is not set: Let AUTOMOC and AUTOUIC process .hh files.  Run
  "cmake --help-policy CMP0100" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  For compatibility, CMake is excluding the header file(s):

    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/hid/UsageWidget.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/WaitDialog.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/Global.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/WindowsDriverDialog.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/WaitDialog.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/hid/UsageWidget.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/DeviceView.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/WindowsDriverDialog.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/hid/CollectionWidget.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/DeviceSelectionListItem.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/Window.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/FlowLayout.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/hid/ControlWidget.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/hid/ItemWidget.hh"
    "/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/hidviz/src/DeviceSelector.hh"

  from processing by AUTOMOC and AUTOUIC.  If any of the files should be
  processed, set CMP0100 to NEW.  If any of the files should not be
  processed, explicitly exclude them by setting the source file property
  SKIP_AUTOGEN:

    set_property(SOURCE file.hh PROPERTY SKIP_AUTOGEN ON)

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/build
==> Sources are ready.
==> Making package: hidviz 0.1.5-1 (Wed 27 Oct 2021 09:35:30 AM CEST)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> WARNING: Using existing $srcdir/ tree
==> Starting build()...
[  2%] Running cpp protocol buffer compiler on src/Buffer.proto
[  5%] Building CXX object libhidx/libhidx_server/CMakeFiles/hidx_server.dir/src/Server.cc.o
[  8%] Building CXX object libhidx/libhidx_server/CMakeFiles/hidx_server.dir/src/Utils.cc.o
[ 11%] Building CXX object libhidx/libhidx_server/CMakeFiles/hidx_server.dir/Buffer.pb.cc.o
[ 13%] Linking CXX static library libhidx_server.a
[ 13%] Built target hidx_server
[ 16%] Building CXX object libhidx/libhidx_server_daemon/CMakeFiles/libhidx_server_daemon.dir/src/main.cc.o
[ 19%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Device.cc.o
[ 22%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/LibHidx.cc.o
[ 25%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/InterfaceHandle.cc.o
[ 27%] Linking CXX executable libhidx_server_daemon
[ 27%] Built target libhidx_server_daemon
[ 30%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Interface.cc.o
[ 33%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Parser.cc.o
[ 36%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Item.cc.o
In file included from /home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:21:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/Interface.hh:76:14: error: ‘shared_ptr’ in namespace ‘std’ does not name a template type
   76 |         std::shared_ptr<InterfaceHandle> getHandle();
      |              ^~~~~~~~~~
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/Interface.hh:29:1: note: ‘std::shared_ptr’ is defined in header ‘<memory>’; did you forget to ‘#include <memory>’?
   28 | #include <string>
  +++ |+#include <memory>
   29 | 
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/Interface.hh:120:14: error: ‘weak_ptr’ in namespace ‘std’ does not name a template type
  120 |         std::weak_ptr<InterfaceHandle> m_handle;
      |              ^~~~~~~~
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/Interface.hh:120:9: note: ‘std::weak_ptr’ is defined in header ‘<memory>’; did you forget to ‘#include <memory>’?
  120 |         std::weak_ptr<InterfaceHandle> m_handle;
      |         ^~~
[ 38%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Collection.cc.o
[ 41%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Control.cc.o
[ 44%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Usages.cc.o
[ 47%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/hid/Usage.cc.o
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc: In member function ‘void libhidx::Interface::parseHidReportDesc()’:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:84:23: error: ‘getHandle’ was not declared in this scope; did you mean ‘handle’?
   84 |         auto handle = getHandle();
      |                       ^~~~~~~~~
      |                       handle
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:95:114: error: no matching function for call to ‘libhidx::ConnectionException::ConnectionException(<brace-enclosed initializer list>)’
   95 |             throw ConnectionException{data.first, "Libusb control transfer failed: " + std::to_string(data.first)};
      |                                                                                                                  ^
In file included from /home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:24:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:37:18: note: candidate: ‘libhidx::ConnectionException::ConnectionException(int, const string&)’
   37 |         explicit ConnectionException(int ev, const std::string& err) : std::system_error{ev, std::generic_category(),  err}{}
      |                  ^~~~~~~~~~~~~~~~~~~
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:37:18: note:   conversion of argument 1 would be ill-formed:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note: candidate: ‘libhidx::ConnectionException::ConnectionException(const libhidx::ConnectionException&)’
   35 |     class ConnectionException : public std::system_error {
      |           ^~~~~~~~~~~~~~~~~~~
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note:   candidate expects 1 argument, 2 provided
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note: candidate: ‘libhidx::ConnectionException::ConnectionException(libhidx::ConnectionException&&)’
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/InterfaceHandle.hh:35:11: note:   candidate expects 1 argument, 2 provided
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc: In member function ‘void libhidx::Interface::runner()’:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:149:23: error: ‘getHandle’ was not declared in this scope; did you mean ‘handle’?
  149 |         auto handle = getHandle();
      |                       ^~~~~~~~~
      |                       handle
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:161:103: error: no matching function for call to ‘std::vector<unsigned char>::vector(<brace-enclosed initializer list>)’
  161 |                 std::vector<unsigned char> data{std::begin(response.data()), std::end(response.data())};
      |                                                                                                       ^
In file included from /usr/include/c++/11.1.0/vector:67,
                 from /usr/include/c++/11.1.0/functional:62,
                 from /home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/Interface.hh:26,
                 from /home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:21:
/usr/include/c++/11.1.0/bits/stl_vector.h:653:9: note: candidate: ‘template<class _InputIterator, class> std::vector<_Tp, _Alloc>::vector(_InputIterator, _InputIterator, const allocator_type&) [with _InputIterator = _InputIterator; <template-parameter-2-2> = <template-parameter-1-2>; _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  653 |         vector(_InputIterator __first, _InputIterator __last,
      |         ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:653:9: note:   template argument deduction/substitution failed:
/usr/include/c++/11.1.0/bits/stl_vector.h:625:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::initializer_list<_Tp>, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  625 |       vector(initializer_list<value_type> __l,
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:625:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11.1.0/bits/stl_vector.h:607:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  607 |       vector(vector&& __rv, const allocator_type& __m)
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:607:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11.1.0/bits/stl_vector.h:589:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&, const allocator_type&, std::false_type) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>; std::false_type = std::integral_constant<bool, false>]’
  589 |       vector(vector&& __rv, const allocator_type& __m, false_type)
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:589:7: note:   candidate expects 3 arguments, 2 provided
/usr/include/c++/11.1.0/bits/stl_vector.h:585:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&, const allocator_type&, std::true_type) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>; std::true_type = std::integral_constant<bool, true>]’
  585 |       vector(vector&& __rv, const allocator_type& __m, true_type) noexcept
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:585:7: note:   candidate expects 3 arguments, 2 provided
/usr/include/c++/11.1.0/bits/stl_vector.h:575:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  575 |       vector(const vector& __x, const allocator_type& __a)
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:575:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11.1.0/bits/stl_vector.h:572:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  572 |       vector(vector&&) noexcept = default;
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:572:7: note:   candidate expects 1 argument, 2 provided
/usr/include/c++/11.1.0/bits/stl_vector.h:553:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  553 |       vector(const vector& __x)
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:553:7: note:   candidate expects 1 argument, 2 provided
/usr/include/c++/11.1.0/bits/stl_vector.h:522:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>::size_type, const value_type&, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::size_type = long unsigned int; std::vector<_Tp, _Alloc>::value_type = unsigned char; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  522 |       vector(size_type __n, const value_type& __value,
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:522:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11.1.0/bits/stl_vector.h:510:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>::size_type, const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::size_type = long unsigned int; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  510 |       vector(size_type __n, const allocator_type& __a = allocator_type())
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:510:7: note:   conversion of argument 1 would be ill-formed:
/usr/include/c++/11.1.0/bits/stl_vector.h:497:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector(const allocator_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<unsigned char>]’
  497 |       vector(const allocator_type& __a) _GLIBCXX_NOEXCEPT
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:497:7: note:   candidate expects 1 argument, 2 provided
/usr/include/c++/11.1.0/bits/stl_vector.h:487:7: note: candidate: ‘std::vector<_Tp, _Alloc>::vector() [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’
  487 |       vector() = default;
      |       ^~~~~~
/usr/include/c++/11.1.0/bits/stl_vector.h:487:7: note:   candidate expects 0 arguments, 2 provided
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc: At global scope:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:176:38: error: no declaration matches ‘std::shared_ptr<libhidx::InterfaceHandle> libhidx::Interface::getHandle()’
  176 |     std::shared_ptr<InterfaceHandle> Interface::getHandle() {
      |                                      ^~~~~~~~~
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:176:38: note: no functions named ‘std::shared_ptr<libhidx::InterfaceHandle> libhidx::Interface::getHandle()’
In file included from /home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:21:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/include/libhidx/Interface.hh:45:11: note: ‘class libhidx::Interface’ defined here
   45 |     class Interface {
      |           ^~~~~~~~~
[ 50%] Building CXX object libhidx/libhidx/CMakeFiles/hidx.dir/src/Connector.cc.o
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc: In member function ‘void libhidx::Interface::sendOutputReport(const std::vector<unsigned char>&)’:
/home/karrq/.cache/yay/hidviz/src/hidviz-0.1.5/libhidx/libhidx/src/Interface.cc:253:23: error: ‘getHandle’ was not declared in this scope; did you mean ‘handle’?
  253 |         auto handle = getHandle();
      |                       ^~~~~~~~~
      |                       handle
make[2]: *** [libhidx/libhidx/CMakeFiles/hidx.dir/build.make:118: libhidx/libhidx/CMakeFiles/hidx.dir/src/Interface.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:222: libhidx/libhidx/CMakeFiles/hidx.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
</details>

EDIT: updating `libhidx` manually to `master` and _then_ building worked

**Comment by ondrejbudai** [comment]
Fixed in 0.2
